File: BEmu/ReferenceDataRequest/MessageReference.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="ReferenceDataRequest/MessageReference.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.ReferenceDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class MessageReference : Message
    {
        private readonly ElementReferenceSecurityDataArray _securities;

        internal MessageReference(CorrelationID corr, Dictionary<string, Dictionary<string, object>> fieldData) : base(new Name("ReferenceDataResponse"), corr, null)
        {
            this._securities = new ElementReferenceSecurityDataArray(fieldData);
        }

        public override int NumElements { get { return 1; } }
        public override IEnumerable<Element> Elements { get { yield return this._securities; } }
        public override string TopicName { get { return ""; } }
        public override BEmu.Element AsElement { get { return new ElementReference(this); } }

        public override object this[string name, int index]
        {
            get
            {
                ElementReferenceSecurityDataArray secArray = (ElementReferenceSecurityDataArray)this[name]; //parent class checks the name

                int itemCount = 0;
                foreach (var item in secArray.Elements)
                {
                    if (itemCount == index)
                        return item;

                    itemCount++;
                }

                return base[name, index];
            }
        }

        public override string ToString()
        {
            StringBuilder result = new StringBuilder();
            result.Append("ReferenceDataResponse (choice) = {" + Environment.NewLine);
            result.Append(this._securities.PrettyPrint(1));
            result.Append("}");

            return result.ToString();
        }

    }
}

File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/ReferenceDataRequest/MessageReference.java
//------------------------------------------------------------------------------
// <copyright project="BEmuJava" file="BEmu.ReferenceDataRequest.MessageReference.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bemu.BEmu.ReferenceDataRequest;

import com.bemu.BEmu.Name;
import com.bemu.BEmu.Element;
import com.bemu.BEmu.Message;
import com.bemu.BEmu.CorrelationID;

import java.util.Map;
import java.util.List;
import java.util.ArrayList;

public class MessageReference extends Message
{
    private final ElementReferenceSecurityDataArray _securities;

    MessageReference(CorrelationID corr, Map<String, Map<String, Object>> fieldData) throws Exception
    {
    	super(new Name("ReferenceDataResponse"), corr, null);
        this._securities = new ElementReferenceSecurityDataArray(fieldData);
    }
    
    public int numElements()
    {
    	return 1;
    }
    
    List<Element> elements()
    {
    	List<Element> result = new ArrayList<Element>();
    	result.add(this._securities);
    	return result;
    }
    
    public Element getElement(String name) throws Exception
    {
    	if(name.equals(this._securities.name().toString()))
    		return this._securities;
    	
    	else
    		return super.getElement(name);
    }
    
    public String topicName()
    {
    	return "";
    }
    
    public Element asElement()
    {
    	return new ElementReference(this);
    }
    
    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
        StringBuilder result = new StringBuilder();

        result.append(String.format("fieldData[] = {%s", System.getProperty("line.separator")));
        result.append(this._securities.prettyPrint(1));
        result.append(String.format("}%s", System.getProperty("line.separator")));
        
        return result;
    }
    
}
